config/deployment.yml

```diff
@@ -1,13 +1,25 @@
 #@ load("@ytt:data", "data")
+#@ load("@ytt:assert", "assert")
 
 apiVersion: apps/v1
 kind: Deployment
 metadata:
   name: secretgen-controller
   namespace: #@ data.values.namespace
   annotations:
-    secretgen-controller.carvel.dev/version: #@ data.values.secretgen_controller_version
+    secretgen-controller.carvel.dev/version: #@ data.values.secretgen_controller_version or assert.fail("'secretgen_controller_version' must be set")
```

Comment:
i tried this `or` thing in the ytt playground 
```

#@ load("@ytt:data", "data")

list_with_a_map: #@ data.values.my_math or 'foo'
```

and just got 
```
- struct has no .my_math field or method (did you mean .my_map?)
    in <toplevel>
      config.yml:6 | list_with_a_map: #@ data.values.my_math or 'foo'
```
      
 are you sure this works?

Reply:
Can you post a link to the playground example you were trying? It does work for me. Looks like maybe you are trying it on a map, whereas version is just a string here. So perhaps that's why you are getting an error?

https://carvel.dev/ytt/#gist:https://gist.github.com/neil-hickey/3005e22ca03588264c05a4cc42bae47c

Reply:
tbh i'm perpetually confused about the playground links,  and the one you sent me doesn't work w/out changing the filename from txt -> yml.  but try this:
```
#@ load("@ytt:assert", "assert")
#@ load("@ytt:data", "data")

#@ p = "123"
#@ f = ""

key1: #@ p or assert.fail("fail")
key2: #@ f or assert.fail("fail")
key3: #@ data.values.foo or assert.fail("fail")
key4: #@ x or assert.fail("fail")
```

in those the p and f work as you want,  but accessing a nonexistent member of the data.values struct triggers a "struct has no ..." before it hits the `or` clause.  (as does asking for a nonexistent variable `x` on the line below)

Reply:
Right, if the variable is missing it will fail on the error you pointed out. I figured that was fine, the purpose of the assertion was to ensure you give a non-empty string, not that you give the variable at all. 

I actually got the pattern from the docs:

https://carvel.dev/ytt/docs/v0.41.0/lang-ref-ytt/#assert

`x = data.values.env.mysql_password or assert.fail("missing env.mysql_password")`